File: crates/kjarni-ffi/bindings/python/example/compare.py
import time
import numpy as np
from kjarni import Embedder, Classifier, version
from sentence_transformers import SentenceTransformer
from sentence_transformers.util import cos_sim

# -------------------------------------------------
# TEST DATA (100+ non-trivial sentences)
# -------------------------------------------------

np.random.seed(42)

topics = [
    "machine learning", "air traffic control", "distributed systems",
    "natural language processing", "computer graphics",
    "financial regulation", "software architecture",
    "real-time systems", "data pipelines", "neural networks"
]

verbs = [
    "optimizes", "evaluates", "analyzes", "processes",
    "transforms", "validates", "predicts", "controls"
]

objects = [
    "large-scale datasets", "complex state transitions",
    "high-dimensional embeddings", "streaming telemetry",
    "regulatory constraints", "fault-tolerant pipelines",
    "user-generated content", "multimodal signals"
]

sentences = [
    f"The {np.random.choice(topics)} system {np.random.choice(verbs)} "
    f"{np.random.choice(objects)} under real-world conditions."
    for _ in range(120)
]

# -------------------------------------------------
# BENCHMARK SETTINGS
# -------------------------------------------------

RUNS = 100
TEXTS = sentences

def benchmark(fn):
    times = []
    for _ in range(RUNS):
        start = time.perf_counter()
        fn()
        times.append(time.perf_counter() - start)
    return np.mean(times), np.std(times)

# -------------------------------------------------
# KJARNI (BATCH)
# -------------------------------------------------

print("\n--- Kjarni Embedder (batch) ---")
kjarni = Embedder(quiet=True, model="minilm-l6-v2")

# Warm-up
_ = kjarni.encode_batch(TEXTS)

def kjarni_batch():
    _ = kjarni.encode_batch(TEXTS)

k_mean, k_std = benchmark(kjarni_batch)

k_embedding = np.array(kjarni.encode("Hello, world!"), dtype=np.float32)
print(f"Kjarni dim: {len(k_embedding)}")
print(f"Kjarni first 5: {k_embedding[:5]}")

k_sim = kjarni.similarity("cat", "dog")
print(f"Kjarni similarity(cat, dog): {k_sim:.6f}")

# -------------------------------------------------
# SENTENCE TRANSFORMERS (BATCH)
# -------------------------------------------------

print("\n--- SentenceTransformers (batch) ---")
model_name = "sentence-transformers/all-MiniLM-L6-v2"
st = SentenceTransformer(model_name)

# Warm-up
_ = st.encode(TEXTS, convert_to_numpy=True, normalize_embeddings=False)

def st_batch():
    _ = st.encode(TEXTS, convert_to_numpy=True, normalize_embeddings=False)

st_mean, st_std = benchmark(st_batch)

st_embedding = st.encode("Hello, world!", convert_to_numpy=True)
print(f"ST dim: {len(st_embedding)}")
print(f"ST first 5: {st_embedding[:5]}")

st_cat = st.encode("cat", convert_to_numpy=True)
st_dog = st.encode("dog", convert_to_numpy=True)
st_sim = cos_sim(st_cat, st_dog).item()
print(f"ST similarity(cat, dog): {st_sim:.6f}")


# -------------------------------------------------
# RESULTS
# -------------------------------------------------

print("\n--- Benchmark Results (100 runs, batch size = 120) ---")
print(f"Kjarni batch encode: {k_mean * 1000:.2f} ms ± {k_std * 1000:.2f} ms")
print(f"ST batch encode:     {st_mean * 1000:.2f} ms ± {st_std * 1000:.2f} ms")

speedup = st_mean / k_mean if k_mean > 0 else float("inf")
print(f"\nRelative speed (ST / Kjarni): {speedup:.2f}x")

File: crates/kjarni-ffi/bindings/csharp/Kjarni.Tests/Compare.cs
using System;
using System.Collections.Generic;
using System.Diagnostics;
using System.Linq;
using Kjarni;

namespace Kjarni.Tests
{
    public static class EmbedderBenchmark
    {
        // -------------------------------------------------
        // TEST DATA (100+ non-trivial sentences)
        // -------------------------------------------------
        
        static readonly string[] Topics = {
            "machine learning", "air traffic control", "distributed systems",
            "natural language processing", "computer graphics",
            "financial regulation", "software architecture",
            "real-time systems", "data pipelines", "neural networks"
        };

        static readonly string[] Verbs = {
            "optimizes", "evaluates", "analyzes", "processes",
            "transforms", "validates", "predicts", "controls"
        };

        static readonly string[] Objects = {
            "large-scale datasets", "complex state transitions",
            "high-dimensional embeddings", "streaming telemetry",
            "regulatory constraints", "fault-tolerant pipelines",
            "user-generated content", "multimodal signals"
        };

        static string[] GenerateSentences(int count, int seed = 42)
        {
            var random = new Random(seed);
            var sentences = new string[count];
            
            for (int i = 0; i < count; i++)
            {
                var topic = Topics[random.Next(Topics.Length)];
                var verb = Verbs[random.Next(Verbs.Length)];
                var obj = Objects[random.Next(Objects.Length)];
                sentences[i] = $"The {topic} system {verb} {obj} under real-world conditions.";
            }
            
            return sentences;
        }

        // -------------------------------------------------
        // BENCHMARK SETTINGS
        // -------------------------------------------------
        
        const int Runs = 10;
        const int SentenceCount = 120;

        static (double Mean, double StdDev) Benchmark(Action action)
        {
            var times = new List<double>();
            
            for (int i = 0; i < Runs; i++)
            {
                var sw = Stopwatch.StartNew();
                action();
                sw.Stop();
                times.Add(sw.Elapsed.TotalSeconds);
            }
            
            var mean = times.Average();
            var stdDev = Math.Sqrt(times.Select(t => Math.Pow(t - mean, 2)).Average());
            
            return (mean, stdDev);
        }

        public static void Run()
        {
            Console.WriteLine();
            Console.WriteLine("============================================================");
            Console.WriteLine("KJARNI C# EMBEDDER BENCHMARK");
            Console.WriteLine("============================================================");
            Console.WriteLine();

            var sentences = GenerateSentences(SentenceCount);
            Console.WriteLine($"Generated {sentences.Length} test sentences");
            Console.WriteLine($"Sample: \"{sentences[0]}\"");
            Console.WriteLine();

            // -------------------------------------------------
            // KJARNI EMBEDDER
            // -------------------------------------------------

            Console.WriteLine("--- Kjarni Embedder (batch) ---");
            
            using var embedder = new Embedder(
                model: "minilm-l6-v2",
                device: "cpu",
                quiet: true
            );

            // Warm-up
            Console.WriteLine("Warming up...");
            _ = embedder.EncodeBatch(sentences);

            // Benchmark batch encoding
            Console.WriteLine($"Running {Runs} iterations...");
            var (mean, stdDev) = Benchmark(() => embedder.EncodeBatch(sentences));

            // Single embedding test
            var embedding = embedder.Encode("Hello, world!");
            Console.WriteLine($"Kjarni dim: {embedding.Length}");
            Console.WriteLine($"Kjarni first 5: [{string.Join(", ", embedding.Take(5).Select(v => v.ToString("F6")))}]");

            // Similarity test
            var similarity = embedder.Similarity("cat", "dog");
            Console.WriteLine($"Kjarni similarity(cat, dog): {similarity:F6}");

            // -------------------------------------------------
            // RESULTS
            // -------------------------------------------------

            Console.WriteLine();
            Console.WriteLine($"--- Benchmark Results ({Runs} runs, batch size = {SentenceCount}) ---");
            Console.WriteLine($"Kjarni batch encode: {mean * 1000:F2} ms ± {stdDev * 1000:F2} ms");
            Console.WriteLine($"Throughput: {SentenceCount / mean:F0} sentences/sec");
            
            // Memory info
            var process = Process.GetCurrentProcess();
            Console.WriteLine($"Working set: {process.WorkingSet64 / (1024 * 1024):F1} MB");

            Console.WriteLine();
            Console.WriteLine("============================================================");
            Console.WriteLine("BENCHMARK COMPLETE");
            Console.WriteLine("============================================================");
        }
    }
}